compiler/src/iree/compiler/Codegen/Common/WorkgroupSpecializationPass.cpp

```diff
@@ -117,101 +68,56 @@ static SmallVector<AffineMinOp> getBoundedTileSizeOps(scf::ForOp forOp,
-//   2. create a condition for scf.if
-//   3. clone the body of the innermost loop in the else block.
-//   4. clone the body of the else block into the then block and update the
-//      bounded size ops with the constant tile size during the cloning.
-static void specializeDistributionLoops(
-    SmallVector<LoopTilingAndDistributionInfo> &infoVec) {
-  // get the scf for loop nests and their tile size.
-  SmallVector<scf::ForOp> distLoops;
-  SmallVector<int64_t> tileSizes;
+// 1. Walk the code and collect affine.min that only depend on workgroup.id
+// and have one constant result.
+// 2. Move those at the top of the function
+// 3. Create a condition that ANDs all the affineMin == constant
+// 4. Splice the rest of the block and clone into a specialized if/else
+static void specializeFunction(func::FuncOp funcOp) {
   SmallVector<AffineMinOp> minSizeOps;
-
-  // Distribution info vector has the innermost loop at index 0.
-  for (auto &info : infoVec) {
-    auto forOp = cast<scf::ForOp>(info.loop);
-    distLoops.push_back(forOp);
-    if (!info.tileSize) {
-      return;
-    }
-    int64_t tileSize = *info.tileSize;
-    tileSizes.push_back(tileSize);
-    SmallVector<AffineMinOp> boundedTileSizesOps =
-        getBoundedTileSizeOps(forOp, tileSize);
-    if (boundedTileSizesOps.empty()) {
-      // When there is no bounded tile size op, it means that the tensor
-      // size is already a multiple of tile size.
-      minSizeOps.push_back(AffineMinOp());
-    } else if (boundedTileSizesOps.size() != 1) {
-      // Supposed to see only one op for the bounded tile size if any.
-      return;
-    } else {
-      minSizeOps.push_back(boundedTileSizesOps[0]);
+  SmallVector<Operation *> ids;
+  funcOp.walk([&minSizeOps, &ids](Operation *op) {
+    if (auto affineMin = dyn_cast<AffineMinOp>(op)) {
+      for (Value operand : affineMin->getOperands()) {
+        if (!operand.getDefiningOp<IREE::HAL::InterfaceWorkgroupIDOp>()) {
+          return WalkResult::advance();
+        }
+        ids.push_back(operand.getDefiningOp());
+      }
+      if (!getConstantLowerBound(affineMin)) {
+        return WalkResult::advance();
+      }
+      minSizeOps.push_back(affineMin);
     }
+    return WalkResult::advance();
+  });
+  if (minSizeOps.empty()) {
+    return;
   }
 
-  // Check the eligilbility. Unsupported cases are:
-  //   1. Dynamic cases
-  //   2. UB < Tile size
-  //   3. UB == a multiple of the tile size
-  bool isAlreadyMultiple = true;
-  for (unsigned i = 0, e = distLoops.size(); i != e; ++i) {
-    scf::ForOp loop = distLoops[i];
-    auto ubOp = loop.getUpperBound().getDefiningOp<arith::ConstantIndexOp>();
-    if (!ubOp) {
-      // TODO: handle dynamic cases by adding more code to check the
-      // conditions.
-      return;
-    }
-    int64_t ub = ubOp.value();
-    int64_t tileSize = tileSizes[i];
-    if (ub % tileSize != 0) {
-      isAlreadyMultiple = false;
-    }
-    if (ub < tileSize) {
```

Comment:
Don't we need these checks in the new code too?

Reply:
I thought there would be no affine_min in this case. Let me double check.

Reply:
checking again, the tileAndDistribute won't create affine_min in this case so nothing happen